pkg/middleware/loggermw/logger.go

```diff
@@ -0,0 +1,174 @@
+)
+
+type Logger interface {
+	Middleware() web.Middleware
+}
+
+type loggerImpl struct {
+	cfg   *setting.Cfg
+	flags featuremgmt.FeatureToggles
+}
+
+func Provide(
+	cfg *setting.Cfg,
+	flags featuremgmt.FeatureToggles,
+) Logger {
+	return &loggerImpl{
+		cfg:   cfg,
+		flags: flags,
+	}
+}
+
+func (l *loggerImpl) Middleware() web.Middleware {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			start := time.Now()
+
+			// we have to init the context with the counter here to update the request
+			r = r.WithContext(log.InitCounter(r.Context()))
+			// put the start time on context so we can measure it later.
+			r = r.WithContext(log.InitstartTime(r.Context(), time.Now()))
+
+			if l.flags.IsEnabled(featuremgmt.FlagUnifiedRequestLog) {
+				r = r.WithContext(errutil.SetUnifiedLogging(r.Context()))
+			}
+
+			rw := web.Rw(w, r)
+			next.ServeHTTP(rw, r)
+
+			duration := time.Since(start)
+			timeTaken := duration / time.Millisecond
+			ctx := contexthandler.FromContext(r.Context())
+			if ctx != nil && ctx.PerfmonTimer != nil {
+				ctx.PerfmonTimer.Observe(float64(timeTaken))
+			}
+
+			if ctx != nil {
+				logParams, logger := l.prepareLogParams(ctx, duration)
+				logger.LogFunc(ctx.Logger)("Request Completed", logParams...)
+			}
+		})
+	}
+}
+
+func (l *loggerImpl) prepareLogParams(c *contextmodel.ReqContext, duration time.Duration) ([]any, errutil.LogLevel) {
+	rw := c.Resp
+	r := c.Req
+
+	status := rw.Status()
+	lvl := errutil.LevelInfo
+
+	switch {
+	case status == http.StatusOK, status == http.StatusNotModified:
+		if !l.cfg.RouterLogging {
+			lvl = errutil.LevelNever
+		}
+	case status >= http.StatusInternalServerError:
+		lvl = errutil.LevelError
+	}
+
+	logParams := []any{
+		"method", r.Method,
+		"path", r.URL.Path,
+		"status", status,
+		"remote_addr", c.RemoteAddr(),
+		"time_ms", int64(duration / time.Millisecond),
+		"duration", duration.String(),
+		"size", rw.Size(),
+	}
+
+	referer, err := SanitizeURL(r.Referer())
+	// We add an empty referer when there's a parsing error, hence this is before the err check.
+	logParams = append(logParams, "referer", referer)
+	if err != nil {
+		logParams = append(logParams, "refererParsingErr", err)
+		lvl = lvl.HighestOf(errutil.LevelWarn)
+	}
+
+	if l.flags.IsEnabled(featuremgmt.FlagDatabaseMetrics) {
+		logParams = append(logParams, "db_call_count", log.TotalDBCallCount(c.Req.Context()))
+	}
+
+	if handler, exist := middleware.RouteOperationName(c.Req); exist {
+		logParams = append(logParams, "handler", handler)
+	}
+
+	logParams = append(logParams, errorLogParams(c.Error)...)
+
+	return logParams, lvl
+}
+
+func errorLogParams(err error) []any {
+	if err == nil {
+		return nil
+	}
+
+	var gfErr errutil.Error
+	if !errors.As(err, &gfErr) {
+		return []any{"err", err.Error()}
+	}
+
+	return []any{
+		"errReason", gfErr.Reason,
+		"errMessageID", gfErr.MessageID,
+		"err", gfErr.LogMessage,
```

Comment:
Nit. Should we use `err` or `error`? We used `error` in response.go earlier. Think both being used throughout the code base.